bee-network/bee-autopeering/src/peering/neighbor.rs

```diff
@@ -0,0 +1,323 @@
+    neighbors: Vec<Neighbor>,
+}
+
+impl<const N: usize, const INBOUND: bool> NeighborhoodInner<N, INBOUND> {
+    // TODO: revisit dead code
+    #[allow(dead_code)]
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    pub(crate) fn insert_neighbor(&mut self, peer: Peer, local: &Local) -> bool {
+        // If the peer already exists remove it.
+        // NOTE: It's a bit less efficient doing it like this, but the code requires less branching this way.
+        let _ = self.remove_neighbor(peer.peer_id());
+
+        if self.neighbors.len() >= N {
+            return false;
+        }
+
+        // Calculate the distance to that peer.
+        let distance = salt_distance(local.read().peer_id(), peer.peer_id(), &{
+            if INBOUND {
+                local.read().private_salt().expect("missing private salt").clone()
+            } else {
+                local.read().public_salt().expect("missing public salt").clone()
+            }
+        });
+
+        self.neighbors.push(Neighbor { distance, peer });
+
+        true
+    }
+
+    pub(crate) fn remove_neighbor(&mut self, peer_id: &PeerId) -> Option<Peer> {
+        if self.neighbors.is_empty() {
+            None
+        } else if let Some(index) = self.neighbors.iter().position(|pd| pd.peer().peer_id() == peer_id) {
+            let Neighbor { peer, .. } = self.neighbors.remove(index);
+            Some(peer)
+        } else {
+            None
+        }
+    }
+
+    /// Check whether the candidate is a suitable neighbor.
+    pub(crate) fn select(&mut self, candidate: Neighbor) -> Option<Peer> {
+        if let Some(furthest) = self.find_furthest() {
+            if &candidate < furthest {
+                Some(candidate.into_peer())
+            } else {
+                None
+            }
+        } else {
+            Some(candidate.into_peer())
+        }
+    }
+
+    /// From the candidate list pick the first, that is closer than the currently furthest neighbor.
+    pub(crate) fn select_from_candidate_list<'a>(&mut self, candidates: &'a [&'a Neighbor]) -> Option<&'a Peer> {
+        if candidates.is_empty() {
+            None
+        } else if let Some(furthest) = self.find_furthest() {
+            for candidate in candidates {
+                if *candidate < furthest {
+                    return Some(candidate.peer());
+                }
+            }
+            None
+        } else {
+            // Any candidate can be selected: pick the first.
+            Some(candidates[0].peer())
+        }
+    }
+
+    pub(crate) fn find_furthest(&mut self) -> Option<&Neighbor> {
+        if self.neighbors.len() >= N {
+            self.neighbors.sort_unstable();
+            self.neighbors.last()
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn remove_furthest(&mut self) -> Option<Peer> {
+        // Note: Both methods require unique access to `self`, so we need to clone the peer id.
```

Comment:
```suggestion
        // Note: Both methods require unique access to `self`, so we need to copy the peer id.
```

Reply:
oh yea, good catch...I forgot to update the docs when I made `PeerId` Copy.

Reply:
Fixed